R/db-odbc-teradata.R

```diff
@@ -0,0 +1,94 @@
+#' @export
+sql_select.Teradata <- function(con, select, from, where = NULL,
+                                             group_by = NULL, having = NULL,
+                                             order_by = NULL,
+                                             limit = NULL,
+                                             distinct = FALSE,
+                                             ...) {
+  out <- vector("list", 7)
+  names(out) <- c("select", "from", "where", "group_by",
+                  "having", "order_by","limit")
+
+  assert_that(is.character(select), length(select) > 0L)
+  out$select    <- build_sql(
+    "SELECT ",
+
+    if (distinct) sql("DISTINCT "),
+
+    # Teradata uses the TOP statement instead of LIMIT which is what SQL92 uses
+    # TOP is expected after DISTINCT and not at the end of the query
+    # e.g: SELECT TOP 100 * FROM my_table
+    if (!is.null(limit) && !identical(limit, Inf)) {
+      assert_that(is.numeric(limit), length(limit) == 1L, limit > 0)
+      build_sql(" TOP ", as.integer(limit), " ")},
+
+    escape(select, collapse = ", ", con = con)
+  )
+
+  out$from      <- sql_clause_from(from, con)
+  out$where     <- sql_clause_where(where, con)
+  out$group_by  <- sql_clause_group_by(group_by, con)
+  out$having    <- sql_clause_having(having, con)
+  out$order_by  <- sql_clause_order_by(order_by, con)
+
+
+  escape(unname(compact(out)), collapse = "\n", parens = FALSE, con = con)
+}
+
+#' @export
+sql_translate_env.Teradata <- function(con) {
+  sql_variant(
+    sql_translator(.parent = base_odbc_scalar,
+      as.numeric    = sql_cast("NUMERIC"),
+      as.double     = sql_cast("NUMERIC"),
+      as.character  = sql_cast("VARCHAR(MAX)"),
+      var           = sql_prefix("VAR_SAMP"),
+      log10         = sql_prefix("LOG"),
+      log           = function(x, base = exp(1)) {
+                        if (isTRUE(all.equal(base, exp(1)))) {
+                          build_sql("ln(", x, ")")
+                        } else {
+                          # Use log change-of-base because postgres doesn't support the
+                          # two-argument "log(base, x)" for floating point x.
+                          build_sql("log(", x, ") / log(", base, ")")
+                        }
+                      },
+      nchar         = sql_prefix("CHARACTER_LENGTH"),
+      ceil          = sql_prefix("CEILING"),
+      ceiling       = sql_prefix("CEILING"),
+      cot           = function(x) build_sql("1 / TAN(", x, ")"),
+      atan2         = function(x, y){
+                          build_sql(
+                            "ATAN2(", y, ",", x, ")"
+                          )},
+      substr        = function(x, start, stop){
+                        len <- stop - start + 1
+                        build_sql(
+                          "SUBSTR(", x, ", ", start, ", ", len, ")"
+                        )},
+      paste         = sql_not_supported("paste()")
```

Comment:
https://forgetcode.com/Teradata/434-CONCAT ?

Reply:
Right, I use `CONCAT` for `paste0`.  `CONCAT` does not add a separator like `paste` does by default.
https://github.com/tidyverse/dbplyr/blob/d08d93071606db414b44f8feb84dae44568c0300/R/translate-sql-odbc.R#L11

Reply:
Ah got it. Might be better to have a more helpful error here? i.e. try `paste0()` instead

Reply:
Good idea, I'd like to create a function for that, can I add `sql_not_supported()` a new `instead` argument? Or should I create a new function?

Reply:
I think you can just create a new function inline